lib/List.dbl

```diff
@@ -36,3 +161,95 @@ pub let iter2 {`re : {type X} -> Unit ->[|_] X} =
     | _                => `re ()
   end
   in iter
+
+pub let init (len : Int) f = 
+  let rec init_aux (i : Int) l f = if i > l then [] else (f i) :: init_aux (i+1) l f 
+  in init_aux 0 (len - 1) f
+
+pub let rec fold_left f acc xs = 
+  match xs with
+  | []      => acc 
+  | x :: xs => fold_left f (f acc x) xs 
+  end 
+
+pub let rec fold_right f xs acc =
+  match xs with 
+  | []      => acc 
+  | x :: xs => f x (fold_right f xs acc)
+  end 
+
+pub let rec for_all p xs =
+  match xs with 
+  | []      => True 
+  | x :: xs => p x && for_all p xs
+  end 
+
+pub let rec exists p xs =
+  match xs with 
+  | []      => False 
+  | x :: xs => p x || exists p xs 
+  end
+
+pub let find {`re : {type X} -> Unit ->[|_] X} p xs = 
+  let rec find_aux p xs =
+    match xs with 
+    | []      => `re () 
+    | x :: xs => if p x then x else find_aux p xs 
+    end  
+  in find_aux p xs
+
+pub let rec find_opt p xs =
+  match xs with 
+  | []      => None 
+  | x :: xs => if p x then Some x else find_opt p xs 
+  end 
+
+pub let rec find_index_opt p xs = 
+  let rec find_index_opt_aux p (i : Int) xs =
+    match xs with 
+    | []      => None 
+    | x :: xs => if p x then Some i else find_index_opt_aux p (i + 1) xs
+    end
+  in find_index_opt_aux p 0 xs
+
+pub let find_map {`re : {type X} -> Unit ->[|_] X} f xs =
+  let rec find_map_aux f xs =
+    match xs with
+    | []      => `re () 
+    | x :: xs => 
+      match f x with
+      | None   => find_map_aux f xs
+      | Some x => x
+      end 
+    end 
+  in find_map_aux f xs
+
+pub let rec find_map_opt f xs =
+  match xs with 
+  | []      => None 
+  | x :: xs => 
+    match f x with 
+    | Some x => Some x
+    | None    => find_map_opt f xs 
+    end 
+  end
+
+pub let rec find_mapi_opt f xs = 
```

Comment:
There is an asymmetry between `find_map` and `find_mapi` functions (the second have only `*_opt` version). Maybe, if they take a parameter that returns `Option`, we could have only the two functions: `find_map` and `find_mapi`, both returning `Option.